File: Doubly_Linked_list.c
#include <stdio.h>
#include <stdlib.h>
struct DLL
{
    int data;
    struct DLL *prev;
    struct DLL *next;
};
struct DLL *First, *Last = NULL;
void display()
{
    struct DLL *temp;
    if (First == NULL)
    {
        printf("List is empty.\n");
    }
    else
    {
        temp = First;
        while (temp != NULL)
        {
            printf("%d <-> ", temp->data);
            temp = temp->next;
        }
        printf("NULL\n");
    }
}
void insert_at_beginning(int element)
{
    struct DLL *NewNode;
    NewNode = (struct DLL *)malloc(sizeof(struct DLL));

    NewNode->data = element;
    NewNode->prev = NULL;
    NewNode->next = NULL;

    if (First == NULL)
    {
        First = Last = NewNode;
    }
    else
    {
        NewNode->next = First;
        First->prev = NewNode;
        First = NewNode;
    }
    printf("%d was inserted at beginning.\n", First->data);
}
void insert_at_end(int element)
{
    struct DLL *NewNode;
    NewNode = (struct DLL *)malloc(sizeof(struct DLL));

    NewNode->data = element;
    NewNode->prev = NULL;
    NewNode->next = NULL;

    if (First == NULL)
    {
        First = Last = NewNode;
    }
    else
    {
        NewNode->prev = Last;
        Last->next = NewNode;
        Last = NewNode;
    }
    printf("%d was inserted at end.\n", Last->data);
}
void insert_at_specificPosition(int element, int POS)
{
    struct DLL *NewNode, *temp, *tempp;
    NewNode = (struct DLL *)malloc(sizeof(struct DLL));

    NewNode->data = element;
    NewNode->prev = NULL;
    NewNode->next = NULL;

    if (First == NULL && POS == 1)
    {
        insert_at_beginning(element);
    }
    else if (POS < 1)
    {
        printf("Invalid position.\n");
        free(NewNode);
    }
    else
    {
        temp = First;
        for (int i = 1; (i < POS - 1) && (temp != NULL); i++)
        {
            temp = temp->next;
        }
        if (temp == NULL) {
            printf("Invalid position: %d cannot be inserted at position %d.\n", NewNode->data, POS);
            free(NewNode);
            return;
        }
        if (temp == Last)
        {
            insert_at_end(element);
        }
        else
        {
            tempp = temp->next;
            NewNode->next = tempp;
            NewNode->prev = temp;
            tempp->prev = NewNode;
            temp->next = NewNode;
            printf("%d was inserted at position %d.\n", NewNode->data, POS);
        }
    }
}
void delete_at_beginning(){

    struct DLL *temp;


    if (First == NULL)
    {
        printf("List is empty.\n");
        return;
    }
    else if (First->next == NULL)
    {
        temp = First;
        First = Last = NULL;
        free(temp);
    }
    else
    {
        temp = First;
        First = First->next;
        First->prev = NULL;
        free(temp);
    }
    printf("Data was deleted from beginning.\n");
}


void delete_at_end(){

    struct DLL *temp;

    if (First == NULL)
    {
        printf("List is empty.\n");
        return;
    }
    else if (First->next == NULL)
    {
        temp = First;
        First = Last = NULL;
        free(temp);
    }
    else
    {
        temp = Last;
        Last = Last->prev;
        Last->next = NULL;
        free(temp);
    }
    printf("Data was deleted from end.\n");
}
void delete_at_specificPosition(int POS)
{
    struct DLL *temp, *tempp;
    if (First == NULL)
    {
        printf("List is empty.\n");
        return;
    }
    else if (POS < 1)
    {
        printf("Invalid position.\n");
        return;
    }
    else if (POS == 1)
    {
        delete_at_beginning();
    }
    else
    {
        temp = First;
        for (int i = 1; (i < POS-1) && (temp->next != NULL); i++)
        
        {
            temp = temp->next;
        }
        if (temp->next == NULL) {
            printf("Data cannot be deleted : Position %d out of bounds.\n",POS);
            return;
        }
        if (temp->next == Last)
        {
            delete_at_end();
        }
        else
        {
            tempp = temp->next;
            temp->next = tempp->next;
            tempp->next->prev = temp;
            free(tempp);
            printf("Data was deleted from position %d.\n", POS);
        }
    }
}

int main()
{
    insert_at_beginning(10);
    display();
    insert_at_end(20);
    display();
    insert_at_specificPosition(15, 2);
    display();
    insert_at_specificPosition(15, 12);
    display();
    delete_at_beginning();
    display();
    delete_at_end();
    display();
    delete_at_specificPosition(7);
    display();
    return 0;
}

File: Doubly_Linked_list.py
class Node:
    def __init__(self, data):  #self-->Refers to the current object being created.
        self.data = data
        self.next = None
        self.prev = None

class DLL:
    def __init__(self):
        self.First = None
        self.Last = None

    def display(self):
        temp = self.First
        if temp is None:
            print("List is empty.")
        else:
            while temp:
                print(temp.data, end="") # end=""--->it continues printing on the same line.
                if temp.next:
                    print(" <-> ", end="")
                temp = temp.next
            print(" <-> NULL")

    def insert_at_beginning(self, element):
        NewNode = Node(element)
        if self.First is None:
            self.First = self.Last = NewNode
        else:
            NewNode.next = self.First;
            self.First.prev = NewNode;
            self.First = NewNode;
        print(f"{self.First.data} was inserted at beginning.")

    def insert_at_end(self, element):
        NewNode = Node(element) #creates a new node object
        if self.First is None:
            self.First = self.Last = NewNode
        else:
            NewNode.prev = self.Last;
            self.Last.next = NewNode;
            self.Last = NewNode;
        print(f"{self.Last.data} was inserted at end.")

    def insert_at_specificPosition(self, element, POS):
        if POS <= 0:
            print("Invalid Position.")
            return
        if POS == 1:
            self.insert_at_beginning(element)
            return
        temp = self.First
        i = 1
        while i < POS - 1 and temp is not None :
            temp = temp.next
            i += 1
        if temp is None:
            NewNode = Node(element)
            print(f"Data {NewNode.data} cannot insert , was out of bounds for position {POS}.")
            return
        if temp == self.Last:
            self.insert_at_end(element)
            return
        NewNode = Node(element)
        tempp = temp.next;
        NewNode.next = tempp;
        NewNode.prev = temp;
        tempp.prev = NewNode;
        temp.next = NewNode;

        print(f"{NewNode.data} was inserted at position {POS}.")

    def delete_at_beginning(self):
        if self.First is None:
            print("List is empty.")
            return
        elif self.First.next is None:
            self.First = self.Last = None
        else:
            temp = self.First;
            self.First = self.First.next;
            self.First.prev = None;
        
        print("Data was deleted from beginning.")

    def delete_at_end(self):
        if self.First is None:
            print("List is empty.")
        elif self.First.next is None:
            self.First = self.Last = None
        else:
            self.Last = self.Last.prev
            self.Last.next = None


        print("Data was deleted from end.")

    def delete_at_specificPosition(self, POS):
        if POS <= 0:
            print("Invalid Deletion.")
            return
        if self.First is None:
            print("List is empty.")
            return
        if POS == 1:
            self.delete_at_beginning()
            return
        temp = self.First
        i = 1
        while i < POS - 1 and temp.next is not None :
            temp = temp.next
            i += 1
        if temp.next is None:
            print(f"Position out of bounds : Data cannot be deleted for position {POS}.")
            return
        if temp.next == self.Last:
            self.delete_at_end()
        else:
            tempp = temp.next
            temp.next = tempp.next
            tempp.next.prev = temp;
        print(f"Data was deleted from the {POS} position.")


dll = DLL() #Make a new empty singly linked list called dll / dll is an object of the dll class
dll.insert_at_beginning(3)
dll.display()
dll.insert_at_end(5)
dll.display()
dll.insert_at_beginning(4)
dll.display()
dll.insert_at_specificPosition(1, 2)
dll.display()
dll.insert_at_specificPosition(15, 6)
dll.display()
dll.delete_at_beginning()
dll.display()
dll.insert_at_specificPosition(9, 3)
dll.display()
dll.delete_at_end()
dll.display()
dll.delete_at_specificPosition(2)
dll.display()
dll.insert_at_specificPosition(7, 8)
dll.display()
dll.delete_at_specificPosition(100)
dll.display()
dll.insert_at_specificPosition(8, 3)
dll.display()